File: cpp/search/bfs.cpp
#include <iostream>
#include <vector>
#include "../../data/graphs/graph.cpp"
#define MAX_NEIGHBORS 4
using namespace std;

//Struct for nodes in graph
struct node{
    int id;
    int n_neighbors = 0;
    struct node *neighbors[MAX_NEIGHBORS];
};

bool in(int val, vector<node *> &v){
    for(int i=0;i<v.size();i++)
        if(v[i]->id == val)
            return true;
    return false;
}

class BFS{
    public:
        int search(node* root, int stop){
            // Nodes on the active edge of search
            vector <node*> active;
            // Nodes that have already been visited
            vector <node*> visited;
            // Nodes to be considered
            vector <node*> next_nodes;

            int n_iterations = 0;
            active.push_back(root);
            //Search until found
            while(1){
                n_iterations++;
                printf("Iteration %d\n", n_iterations);
                //For all activate nodes
                for(int a=0;a<active.size();a++){
                    printf("\nActive node: %d, has %d neighbors:", active[a]->id, active[a]->n_neighbors);
                    // Check neighbors
                    for(int n=0;n<active[a]->n_neighbors;n++){
                        node* neighbor = active[a]->neighbors[n];
                        int id = neighbor->id;
                        printf("%d,", id);
                        // If not already active or visited
                        if(!in(id, active) && !in(id, visited))
                            next_nodes.push_back(neighbor);
                    }
                }
                printf("\n");
                // Check to see if answer was in new nodes
                for(int i=0;i<next_nodes.size();i++)
                    if(next_nodes[i]->id == stop)
                        return n_iterations;
                // Else get ready for next round
                for(int i=0;i<active.size();i++)
                    visited.push_back(active[i]);
                active.clear();
                for(int i=0;i<next_nodes.size();i++)
                    active.push_back(next_nodes[i]);
                next_nodes.clear();
                cin.get();
            }

            return 0;
        }
};

int main(int argc, char **argv){
    node* nodes = new node[51];
    for(int i=0;i<51;i++)
        nodes[i].id = i;
    //Set up connections
    for(int i=0;i<51;i++){
        for(int j=0;j<MAX_NEIGHBORS;j++){
            if(graph[i][j] >= 0){
                nodes[i].n_neighbors = j+1;
                node* neighbor = &(nodes[graph[i][j]]);
                nodes[i].neighbors[j] = &(nodes[graph[i][j]]);
            }
        }
    }
    BFS bfs;
    int s = bfs.search(&(nodes[0]), 15);
    printf("Search took %d iterations",s);
    return s;
}

File: python/search/bfs.py
import datetime


graph1 = {
    0: (5,29,4,7),
    1: (7,),
    2: (18,10),
    3: (9,7),
    4: (0,8),
    5: (0,9),
    6: (38,20),
    7: (0,3,1),
    8: (4,19,38),
    9: (5,22,21,3),
    10: (2,31,32,19),
    11: (26,12,13),
    12: (11,20),
    13: (11,28,27,14),
    14: (13,15),
    15: (14,),
    16: (21,26),
    17: (38,40,37),
    18: (30,2,22),
    19: (10,41,8),
    20: (6,12),
    21: (9,16),
    22: (18,9),
    23: (42,40),
    24: (27,46,33),
    25: (33,34),
    26: (16,11),
    27: (13,24),
    28: (13,),
    29: (0,),
    30: (47,31,18),
    31: (30,44,10),
    32: (44,10,41,45),
    33: (24,25),
    34: (25,36),
    35: (49,50),
    36: (39,34,51),
    37: (17,39),
    38: (8,17,6),
    39: (37,36),
    40: (23,17),
    41: (32,19,45,42),
    42: (41,23),
    43: (46,49),
    44: (47,31,48,32),
    45: (48,32,41),
    46: (24,43),
    47: (30,44),
    48: (44,45),
    49: (43,35),
    50: (35,51),
    51: (50,36)
}

class BFS():
    def __init__(self, graph):
        # A graph is a dict with {node: (neighbors,)}
        self.graph = graph

    def search(self, start, stop):
        if start == stop:
            return 0

        # Leading edge nodes
        active = [start]
        # Nodes that have already been visited
        visited = [start]
        # Record how many iterations it takes
        n_iterations = 0
        # Record how many nodes searched
        n_nodes = 0
        # Keep searching till it's found
        while True:
            n_iterations += 1
            # Get neighbors of active nodes
            neighbors = []
            for a in active:
                # For each of their neighbors
                for n in self.graph[a]:
                    if (n not in active) and (n not in visited) and (n not in neighbors):
                        neighbors.append(n)
                        n_nodes += 1
            # Check if it was found
            if stop in neighbors:
                return n_iterations, n_nodes
            # Set new neighbors to new active and add old active to visited
            for a in active:
                visited.append(a)
            active = neighbors
            
bfs = BFS(graph1)
start = datetime.datetime.now()
n_iterations, n_nodes = bfs.search(0, 15)
stop = datetime.datetime.now()

latency = stop - start
print("Search took",n_iterations,"iterations")
print('Search took',latency.microseconds,'us')